pkg/apis/cluster/validation/validation.go

```diff
@@ -0,0 +1,172 @@
+		}
+		if spec.ImpersonatorSecretRef.Name == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("impersonatorSecretRef").Child("name"), ""))
+		}
+	}
+
+	if spec.ProxyURL != "" {
+		allErrs = append(allErrs, ValidateClusterProxyURL(fldPath.Child("proxyURL"), spec.ProxyURL)...)
+	}
+
+	if len(spec.Taints) > 0 {
+		allErrs = append(allErrs, ValidateClusterTaints(spec.Taints, fldPath.Child("taints"))...)
+	}
+
+	return allErrs
+}
+
+// ValidateClusterAPIEndpoint validates cluster's apiEndpoint
+func ValidateClusterAPIEndpoint(fldPath *field.Path, apiEndpoint string, forceHTTPS bool) field.ErrorList {
+	var allErrs field.ErrorList
+	const form = "; desired format: hostname, hostname:port, IP or IP:port"
+	if u, err := url.Parse(apiEndpoint); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, apiEndpoint, "apiEndpoint must be a valid URL: "+err.Error()+form))
+	} else {
+		if forceHTTPS && u.Scheme != "https" {
+			allErrs = append(allErrs, field.Invalid(fldPath, u.Scheme, "'https' is the only allowed URL scheme"+form))
+		}
+		if len(u.Host) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath, u.Host, "host must be provided"+form))
+		}
+		if u.User != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, u.User.String(), "user information is not permitted in the URL"))
+		}
+		if len(u.Fragment) != 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath, u.Fragment, "fragments are not permitted in the URL"))
+		}
+		if len(u.RawQuery) != 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath, u.RawQuery, "query parameters are not permitted in the URL"))
+		}
+	}
+	return allErrs
+}
+
+// ValidateClusterProxyURL validates cluster's proxyURL.
+func ValidateClusterProxyURL(fldPath *field.Path, proxyURL string) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for _, msg := range utilvalidation.ValidateClusterProxyURL(proxyURL) {
+		allErrs = append(allErrs, field.Invalid(fldPath, proxyURL, msg))
+	}
+	return allErrs
+}
+
+// ValidateClusterTaints tests if given taints have valid data.
+func ValidateClusterTaints(taints []corev1.Taint, fldPath *field.Path) field.ErrorList {
+	allErrors := field.ErrorList{}
+
+	uniqueTaints := map[corev1.TaintEffect]sets.String{}
+
+	for i, currTaint := range taints {
+		idxPath := fldPath.Index(i)
+		// validate the taint key
+		allErrors = append(allErrors, unversionedvalidation.ValidateLabelName(currTaint.Key, idxPath.Child("key"))...)
+		// validate the taint value
+		if errs := validation.IsValidLabelValue(currTaint.Value); len(errs) != 0 {
+			allErrors = append(allErrors, field.Invalid(idxPath.Child("value"), currTaint.Value, strings.Join(errs, ";")))
+		}
+		// validate the taint effect
+		allErrors = append(allErrors, validateTaintEffect(&currTaint.Effect, false, idxPath.Child("effect"))...)
+
+		// validate if taint is unique by <key, effect>
+		if len(uniqueTaints[currTaint.Effect]) > 0 && uniqueTaints[currTaint.Effect].Has(currTaint.Key) {
```

Comment:
```suggestion
		if uniqueTaints[currTaint.Effect].Has(currTaint.Key) {
```
This will be enough, right? 
Don't know if `len(sets.String)` will work.

Reply:
@RainbowMango it is necessary to avoid panic

Reply:
Really? Do you mean if we invoke `Has()` method on an empty `sets.String{}` will panic?

Reply:
> 1. I suggest just putting `validation.go` to `pkg/apis/cluster` directory as it is only for `internel` types.

@RainbowMango

In [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes/tree/master/pkg/apis/apps) and [openshift/origin](https://github.com/openshift/origin/tree/v3.11.0/pkg/apps/apis/apps) repo, the validation files are placed in a separate directory. 

I recommend keeping the same code style as them, what do you think?

Reply:
> I recommend keeping the same code style as them, what do you think?

OK for me.

Reply:
> Really? Do you mean if we invoke `Has()` method on an empty `sets.String{}` will panic?

It may not have been initialized when we invoke `Has()` method.

```go
// ValidateClusterTaints tests if given taints have valid data.
func ValidateClusterTaints(taints []corev1.Taint, fldPath *field.Path) field.ErrorList {
	allErrors := field.ErrorList{}

	uniqueTaints := map[corev1.TaintEffect]sets.String{}
	for i, currTaint := range taints {
             ...
		// validate if taint is unique by <key, effect>
		if len(uniqueTaints[currTaint.Effect]) > 0 && uniqueTaints[currTaint.Effect].Has(currTaint.Key) {
                      ....
		}

		// add taint to existingTaints for uniqueness check
		if len(uniqueTaints[currTaint.Effect]) == 0 {
			uniqueTaints[currTaint.Effect] = sets.String{}
		}
		uniqueTaints[currTaint.Effect].Insert(currTaint.Key)
	}
	return allErrors
}
```

Reply:
@RainbowMango 

`ValidateClusterTaints`  is lifted from [kubernetes validateNodeTaints](https://github.com/kubernetes/kubernetes/blob/885961cda8875cce192700fe6a32ff57ba706106/pkg/apis/core/validation/validation.go#L5003)

Reply:
But, I'm still interest if the `len(uniqueTaints[currTaint.Effect]) > 0` could be removed, though. 
Because, the `sets` has been initilized by `uniqueTaints := map[corev1.TaintEffect]sets.String{}`.